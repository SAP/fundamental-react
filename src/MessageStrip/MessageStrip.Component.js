import { MessageStrip } from '..';
import path from 'path';
import React from 'react';
import { ComponentPage, Example } from '../_playground';

export const MessageStripComponent = () => {
    return (
        <ComponentPage
            description='An **MessageStrip** provides a message within the application that is color-coded to emphasize the level of urgency.'
            sourceModulePath={path.join(__dirname, './MessageStrip')}
            title='MessageStrip'>

            <Example
                description={`The MessageStrip provides information that is useful and relevant, but not critical. It can also provide
                    feedback that an action has been executed. The user will need to dismiss the message.`}
                title='Default MessageStrip'>
                <MessageStrip
                    dismissible
                    link='#'
                    linkText='link'>
                    Default MessageStrip with a{' '}
                </MessageStrip>
            </Example>

            <Example
                description='The MessageStrip warns of potential issues, but the user can still continue. The user will need to dismiss the message.'
                title='Warning MessageStrip'>
                <MessageStrip dismissible type='warning'>
                    Warning Message.
                </MessageStrip>
                <MessageStrip dismissible noGlyph
                    type='warning'>
                    Warning Message with no icon.
                </MessageStrip>
            </Example>

            <Example
                description={`This MessageStrip is triggered after the user entered data incorrectly or a system error has occurred. It
                    should interrupt the user. A final action such as Submit cannot be carried out until the user has
                    rectified the error. The user will need to dismiss the message.`}
                title='Error MessageStrip'>
                <MessageStrip dismissible
                    link='#'
                    linkText='Learn More'
                    type='error'>
                    Error Message.{' '}
                </MessageStrip>
                <MessageStrip dismissible
                    link='#'
                    linkText='Learn More'
                    noGlyph
                    type='error'>
                    Error Message with no icon.{' '}
                </MessageStrip>
            </Example>

            <Example
                title='Success MessageStrip'>
                <MessageStrip dismissible
                    link='#'
                    linkText='Learn More'
                    type='success'>
                    Message Success.{' '}
                </MessageStrip>
                <MessageStrip dismissible
                    link='#'
                    linkText='Learn More'
                    noGlyph
                    type='success'>
                    Message Success with no icon.{' '}
                </MessageStrip>
            </Example>

            <Example
                title='Information MessageStrip'>
                <MessageStrip dismissible
                    link='#'
                    linkText='Learn More'
                    type='information'>
                    Information Message.{' '}
                </MessageStrip>

                <MessageStrip dismissible
                    link='#'
                    linkText='Learn More'
                    noGlyph
                    type='information'>
                    Information Message with no icon.{' '}
                </MessageStrip>
            </Example>
        </ComponentPage>
    );
};
