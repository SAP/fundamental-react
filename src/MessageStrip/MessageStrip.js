import Button from '../Button/Button';
import classnamesBind from 'classnames/bind';
import CustomPropTypes from '../utils/CustomPropTypes/CustomPropTypes';
import Link from '../Link/Link';
import { MESSAGESTRIP_TYPES } from '../utils/constants';
import PropTypes from 'prop-types';
import useUniqueId from '../utils/useUniqueId';
import withStyles from '../utils/withStyles';
import React, { useState } from 'react';
import iconStyles from 'fundamental-styles/dist/icon.css';
import messageStripStyles from 'fundamental-styles/dist/message-strip.css';

const classnames = classnamesBind.bind({
    ...iconStyles,
    ...messageStripStyles
});

/** A **MessageStrip** provides a message within the application that is
 * color-coded to emphasize the level of urgency. */

const MessageStrip = (props) => {

    let [active, setActive] = useState(true);

    const {
        onCloseClicked,
        buttonProps,
        type,
        link,
        linkProps,
        linkText,
        localizedText,
        noGlyph,
        dismissible,
        children,
        className,
        cssNamespace,
        ...otherProps
    } = props;

    const closeMessageStripHandler = (e) => {
        setActive(false);
        onCloseClicked(e);
    };


    const MessageStripClasses = classnames(
        `${cssNamespace}-message-strip`,
        {
            [`${cssNamespace}-message-strip--dismissible`]: dismissible,
            [`${cssNamespace}-message-strip--no-icon`]: noGlyph,
            [`${cssNamespace}-message-strip--${type}`]: !!type
        },
        className
    );

    const generatedAlertId = useUniqueId();
    const alertId = otherProps?.id || generatedAlertId;

    return (
        <>
            {active && (
                <div
                    {...otherProps}
                    className={MessageStripClasses}
                    id={alertId}
                    role='alert'>
                    {dismissible && (
                        <Button
                            {...buttonProps}
                            aria-controls={alertId}
                            aria-label={localizedText.close}
                            className={classnames(`${cssNamespace}-message-strip__close`)}
                            compact
                            glyph='decline'
                            onClick={closeMessageStripHandler}
                            option='transparent' />
                    )}
                    <p className={classnames(`${cssNamespace}-message-strip__text`)}>
                        {children}
                        {link && (
                            <Link
                                {...linkProps}
                                href={link}>
                                {linkText}{' '}
                            </Link>
                        )}
                    </p>
                </div>
            )}
        </>
    );
};

MessageStrip.displayName = 'MessageStrip';

MessageStrip.propTypes = {
    /** Additional props to be spread to the `<button>` element */
    buttonProps: PropTypes.object,
    /** Node(s) to render within the component */
    children: PropTypes.node,
    /** CSS class(es) to add to the element */
    className: PropTypes.string,
    /** Set to **true** to show a dismiss button */
    dismissible: PropTypes.bool,
    /** Value to be applied to the anchor\'s `href` attribute */
    link: PropTypes.string,
    /** Additional props to be spread to the link\'s `<a>` element */
    linkProps: PropTypes.object,
    /** Localized display text of the link */
    linkText: PropTypes.string,
    /** Localized text to be updated based on location/language */
    localizedText: CustomPropTypes.i18n({
        /** Value for aria-label on the close <button> element */
        close: PropTypes.string
    }),
    /** Set to **true** to disable the state icon */
    noGlyph: PropTypes.bool,
    /** Sets the variation of the component. Primarily used for styling:
    'warning',
    'error',
    'success',
    'information'*/
    type: PropTypes.oneOf(MESSAGESTRIP_TYPES),
    /**
     * Callback function; triggered when a dismissible MessageStrip's close button is clicked.
     *
     * @param {SyntheticEvent} event - React's original SyntheticEvent. See https://reactjs.org/docs/events.html.
     * @returns {void}
    */
    onCloseClicked: PropTypes.func
};

MessageStrip.defaultProps = {
    localizedText: {
        close: 'Close'
    },
    onCloseClicked: () => { }
};

export default withStyles(MessageStrip);
