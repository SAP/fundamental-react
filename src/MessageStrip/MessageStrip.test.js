import MessageStrip from './MessageStrip';
import { mount } from 'enzyme';
import React from 'react';

describe('<MessageStrip />', () => {
    describe('Prop spreading', () => {
        test('should allow props to be spread to the MessageStrip component', () => {
            const element = mount(<MessageStrip data-sample='Sample' />);

            expect(
                element.find('div').getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        test('should allow props to be spread to the MessageStrip component\'s button element when dismissible', () => {
            const element = mount(<MessageStrip buttonProps={{ 'data-sample': 'Sample' }} dismissible />);

            expect(
                element.find('button').getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        test('should allow props to be spread to the MessageStrip component\'s a element when link provided', () => {
            const element = mount(<MessageStrip link='#' linkProps={{ 'data-sample': 'Sample' }} />);

            expect(
                element.find('a').getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });
    });

    describe('onClick handler', () => {
        test('should dispatch the onClick callback with the event', () => {
            let f = jest.fn();
            const element = mount(<MessageStrip data-sample='Sample' dismissible
                onCloseClicked={f} />);

            element.find('button').simulate('click');

            expect(f).toHaveBeenCalledTimes(1);
        });
    });
});
