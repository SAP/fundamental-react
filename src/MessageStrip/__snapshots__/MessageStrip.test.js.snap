// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageStrip /> create basic MessageStrip 1`] = `
<div>
  <div
    className="fd-message-strip fd-message-strip--dismissible"
    role="alert"
  >
    <button
      aria-controls="j2ALl423"
      aria-label="Close"
      className="fd-button fd-button--transparent fd-button--compact fd-message-strip__close"
      onClick={[Function]}
      type="button"
    />
    <p
      className="fd-message-strip__text"
    >
      Default MessageStrip with a
      <a
        className="fd-link"
        href="#"
      >
        link
         
      </a>
    </p>
  </div>
</div>
`;

exports[`<MessageStrip /> create basic MessageStrip 2`] = `
<div>
  <div
    className="fd-message-strip fd-message-strip--dismissible fd-message-strip--error"
    role="alert"
  >
    <button
      aria-controls="j2ALl423"
      aria-label="Close"
      className="fd-button fd-button--transparent fd-button--compact fd-message-strip__close"
      onClick={[Function]}
      type="button"
    />
    <p
      className="fd-message-strip__text"
    >
      Error message with a
    </p>
  </div>
</div>
`;

exports[`<MessageStrip /> create non-dismissible MessageStrip 1`] = `
<div>
  <div
    className="fd-message-strip blue"
    role="alert"
  >
    <p
      className="fd-message-strip__text"
    >
      Default MessageStrip that cannot be dismissed
      <a
        className="fd-link"
        href="#"
      >
        link
         
      </a>
    </p>
  </div>
</div>
`;
